kv/kvnemesis: TestKVNemesisSingleNode failed

kv/kvnemesis.TestKVNemesisSingleNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9828507?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9828507?buildTab=artifacts#/) on master @ [5b44c4fd6d97dc3a75e786fed24f47ad8bcf9039](https://github.com/cockroachdb/cockroach/commits/5b44c4fd6d97dc3a75e786fed24f47ad8bcf9039):


```
=== RUN   TestKVNemesisSingleNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode2646109487
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 9116470825307397985
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis85842970
    kvnemesis.go:165: range 113 (/Table/100/"8202395e8dbe1792") RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1682591416804502043 IntentAge:0 GCBytesAge:86808 LiveBytes:2703 LiveCount:51 KeyBytes:8526 KeyCount:186 ValBytes:5845 ValCount:385 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:9 RangeKeyBytes:468 RangeValCount:10 RangeValBytes:107 SysBytes:6491 SysCount:37 AbortSpanBytes:1810}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1682591416804502043 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:-6 SysCount:0 AbortSpanBytes:0}
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis85842970/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis85842970/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis85842970/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis85842970/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:148
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisSingleNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode2646109487
--- FAIL: TestKVNemesisSingleNode (22.58s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27429

Epic CRDB-27234

--- a/pkg/kv/kvserver/batcheval/cmd_end_transaction.go
+++ b/pkg/kv/kvserver/batcheval/cmd_end_transaction.go
@@ -1012,13 +1012,29 @@ func splitTrigger(
 			"unable to compute range key stats delta for RHS")
 	}
 
+	// Retrieve MVCC Stats from the current batch instead of using stats from
+	// execution context. Stats in the context could diverge from storage snapshot
+	// of current request when lease extensions are applied. Lease expiration is
+	// a special case that updates stats without obtaining latches and thus can
+	// execute concurrently with splitTrigger. As a result we must not write
+	// absolute stats values for LHS based on this value, always produce a delta
+	// since underlying stats in storage could change. At the same time it is safe
+	// to write absolute RHS side stats since we hold lock for values, and
+	// "unprotected" lease key don't yet exist until this split operation creates
+	// RHS replica.
+	currentStats, err := MakeStateLoader(rec).LoadMVCCStats(ctx, batch)
+	if err != nil {
+		return enginepb.MVCCStats{}, result.Result{}, errors.Wrap(err,
+			"unable to fetch original range mvcc stats for split")
+	}
+
 	h := splitStatsHelperInput{
-		AbsPreSplitBothEstimated: rec.GetMVCCStats(),
-		DeltaBatchEstimated:      bothDeltaMS,
-		DeltaRangeKey:            rangeKeyDeltaMS,
-		AbsPostSplitLeftFn:       makeScanStatsFn(ctx, batch, ts, &split.LeftDesc, "left hand side"),
-		AbsPostSplitRightFn:      makeScanStatsFn(ctx, batch, ts, &split.RightDesc, "right hand side"),
-		ScanRightFirst:           splitScansRightForStatsFirst || emptyRHS,
+		AbsPreSplitBothStored: currentStats,
+		DeltaBatchEstimated:   bothDeltaMS,
+		DeltaRangeKey:         rangeKeyDeltaMS,
+		PostSplitScanLeftFn:   makeScanStatsFn(ctx, batch, ts, &split.LeftDesc, "left hand side"),
+		PostSplitScanRightFn:  makeScanStatsFn(ctx, batch, ts, &split.RightDesc, "right hand side"),
+		ScanRightFirst:        splitScansRightForStatsFirst || emptyRHS,
 	}
 	return splitTriggerHelper(ctx, rec, batch, h, split, ts)
 }

--- a/pkg/kv/kvserver/batcheval/split_stats_helper.go
+++ b/pkg/kv/kvserver/batcheval/split_stats_helper.go
@@ -16,7 +16,7 @@ import "github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 // split. The quantities known during a split (i.e. while the split trigger
 // is evaluating) are
 //
-//   - AbsPreSplitBothEstimated: the stats of the range before the split trigger,
+//   - AbsPreSplitBothStored: the stats of the range before the split trigger,
 //     i.e. without accounting for any writes in the batch. This can have
 //     ContainsEstimates set.
 //   - DeltaBatchEstimated: the writes in the batch, i.e. the stats delta accrued
@@ -84,14 +84,14 @@ import "github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 // These two equations are easily solved for the unknowns. First, we can express
 // DeltaPostSplitLeft() in known quantities via (2) as
 //
-//	DeltaPostSplitLeft() = AbsPostSplitLeft - AbsPreSplitBothEstimated.
+//	DeltaPostSplitLeft() = AbsPostSplitLeft - AbsPreSplitBothStored.
 //
 // Note that if we start out with estimates, DeltaPostSplitLeft() will wipe out
 // those estimates when added to the absolute stats.
 //
 // For AbsPostSplitRight(), there are two cases. First, due to the identity
 //
-//	CombinedErrorDelta = AbsPreSplitBothEstimated + DeltaBatchEstimated
+//	CombinedErrorDelta = AbsPreSplitBothStored + DeltaBatchEstimated
 //	                     -(AbsPostSplitLeft + AbsPostSplitRight)
 //	                     + DeltaRangeKey.
 //
@@ -119,15 +119,15 @@ type splitStatsScanFn func() (enginepb.MVCCStats, error)
 
 // splitStatsHelperInput is passed to makeSplitStatsHelper.
 type splitStatsHelperInput struct {
-	AbsPreSplitBothEstimated enginepb.MVCCStats
-	DeltaBatchEstimated      enginepb.MVCCStats
-	DeltaRangeKey            enginepb.MVCCStats
-	// AbsPostSplitLeftFn returns the stats for the left hand side of the
-	// split.
-	AbsPostSplitLeftFn splitStatsScanFn
-	// AbsPostSplitRightFn returns the stats for the right hand side of the
-	// split.
-	AbsPostSplitRightFn splitStatsScanFn
+	AbsPreSplitBothStored enginepb.MVCCStats
+	DeltaBatchEstimated   enginepb.MVCCStats
+	DeltaRangeKey         enginepb.MVCCStats
+	// PostSplitScanLeftFn returns the stats for the left hand side of the
+	// split computed by scanning relevant part of the range.
+	PostSplitScanLeftFn splitStatsScanFn
+	// PostSplitScanRightFn returns the stats for the right hand side of the
+	// split computed by scanning relevant part of the range.
+	PostSplitScanRightFn splitStatsScanFn
 	// ScanRightFirst controls whether the left hand side or the right hand
 	// side of the split is scanned first. In cases where neither of the
 	// input stats contain estimates, this is the only side that needs to
@@ -137,7 +137,7 @@ type splitStatsHelperInput struct {
 
 // makeSplitStatsHelper initializes a splitStatsHelper. The values in the input
 // are assumed to not change outside of the helper and must no longer be used.
-// The provided AbsPostSplitLeftFn and AbsPostSplitRightFn recompute the left
+// The provided PostSplitScanLeftFn and PostSplitScanRightFn recompute the left
 // and right hand sides of the split after accounting for the split trigger
 // batch. Each are only invoked at most once, and only when necessary.
 func makeSplitStatsHelper(input splitStatsHelperInput) (splitStatsHelper, error) {
@@ -149,21 +149,21 @@ func makeSplitStatsHelper(input splitStatsHelperInput) (splitStatsHelper, error)
 	var absPostSplitFirst enginepb.MVCCStats
 	var err error
 	if h.in.ScanRightFirst {
-		absPostSplitFirst, err = input.AbsPostSplitRightFn()
+		absPostSplitFirst, err = input.PostSplitScanRightFn()
 		h.absPostSplitRight = &absPostSplitFirst
 	} else {
-		absPostSplitFirst, err = input.AbsPostSplitLeftFn()
+		absPostSplitFirst, err = input.PostSplitScanLeftFn()
 		h.absPostSplitLeft = &absPostSplitFirst
 	}
 	if err != nil {
 		return splitStatsHelper{}, err
 	}
 
-	if h.in.AbsPreSplitBothEstimated.ContainsEstimates == 0 &&
+	if h.in.AbsPreSplitBothStored.ContainsEstimates == 0 &&
 		h.in.DeltaBatchEstimated.ContainsEstimates == 0 {
 		// We have CombinedErrorDelta zero, so use arithmetic to compute the
 		// stats for the second side.
-		ms := h.in.AbsPreSplitBothEstimated
+		ms := h.in.AbsPreSplitBothStored
 		ms.Subtract(absPostSplitFirst)
 		ms.Add(h.in.DeltaBatchEstimated)
 		ms.Add(h.in.DeltaRangeKey)
@@ -177,15 +177,15 @@ func makeSplitStatsHelper(input splitStatsHelperInput) (splitStatsHelper, error)
 
 	// Estimates are contained in the input, so ask the oracle to scan to compute
 	// the stats for the second side. We only scan the second side when either of
-	// the input stats above (AbsPreSplitBothEstimated or DeltaBatchEstimated)
+	// the input stats above (AbsPreSplitBothStored or DeltaBatchEstimated)
 	// contains estimates, so that we can guarantee that the post-splits stats
 	// don't.
 	var absPostSplitSecond enginepb.MVCCStats
 	if h.in.ScanRightFirst {
-		absPostSplitSecond, err = input.AbsPostSplitLeftFn()
+		absPostSplitSecond, err = input.PostSplitScanLeftFn()
 		h.absPostSplitLeft = &absPostSplitSecond
 	} else {
-		absPostSplitSecond, err = input.AbsPostSplitRightFn()
+		absPostSplitSecond, err = input.PostSplitScanRightFn()
 		h.absPostSplitRight = &absPostSplitSecond
 	}
 	if err != nil {
@@ -214,7 +214,7 @@ func (h splitStatsHelper) DeltaPostSplitLeft() enginepb.MVCCStats {
 	// the helper, we can make that work, too.
 	// NB: note how none of this depends on mutations to absPostSplitRight.
 	ms := *h.absPostSplitLeft
-	ms.Subtract(h.in.AbsPreSplitBothEstimated)
+	ms.Subtract(h.in.AbsPreSplitBothStored)
 
 	return ms
 }

